packages/integration-sdk-runtime/src/synchronization/index.ts

```diff
@@ -305,22 +314,103 @@ export async function uploadData<T extends UploadDataLookup, K extends keyof T>(
   data: T[K][],
   uploadBatchSize?: number,
 ) {
-  const batches = chunk(data, uploadBatchSize || UPLOAD_BATCH_SIZE);
-  await pMap(
-    batches,
-    async (batch: T[K][]) => {
-      if (batch.length) {
-        await uploadDataChunk({
-          apiClient,
-          logger,
-          jobId: job.id,
-          type,
-          batch,
-        });
+  let retryAfterTruncate: Boolean;
+  do {
+    // Set/Reset retry flag
+    retryAfterTruncate = false;
+    try {
+      const batches = chunk(data, uploadBatchSize || UPLOAD_BATCH_SIZE);
+      await pMap(
+        batches,
+        async (batch: T[K][]) => {
+          if (batch.length) {
+            await uploadDataChunk({
+              apiClient,
+              logger,
+              jobId: job.id,
+              type,
+              batch,
+            });
+          }
+        },
+        { concurrency: UPLOAD_CONCURRENCY },
+      );
+    } catch (err) {
+      if (
+        err.code === 'RequestEntityTooLargeException' ||
+        err.response?.status === 413
+      ) {
+        logger.info(`Attempting to shrink raw_data`);
+        shrinkRawData(data);
```

Comment:
I wonder how much more efficient this would be if we only pass the actual batch that caused the failure rather than all off the data.